staging/src/k8s.io/apiserver/pkg/server/config_test.go

```diff
@@ -241,3 +250,100 @@ func checkExpectedPathsAtRoot(url string, expectedPaths []string, t *testing.T)
 		}
 	})
 }
+
+func TestAuthenticationAuditAnnotationsDefaultChain(t *testing.T) {
+	authn := authenticator.RequestFunc(func(req *http.Request) (*authenticator.Response, bool, error) {
+		// confirm that we can set an audit annotation in a handler before WithAudit
+		audit.AddAuditAnnotation(req.Context(), "pandas", "are awesome")
+
+		// confirm that trying to use the audit event directly would never work
+		if ae := request.AuditEventFrom(req.Context()); ae != nil {
+			t.Errorf("expected nil audit event, got %v", ae)
+		}
+
+		return &authenticator.Response{User: &user.DefaultInfo{}}, true, nil
+	})
+	backend := &testBackend{}
+	c := &Config{
+		Authentication:     AuthenticationInfo{Authenticator: authn},
+		AuditBackend:       backend,
+		AuditPolicyChecker: policy.FakeChecker(auditinternal.LevelMetadata, nil),
+
+		// avoid nil panics
+		HandlerChainWaitGroup: &waitgroup.SafeWaitGroup{},
+		RequestInfoResolver:   &request.RequestInfoFactory{},
+		RequestTimeout:        10 * time.Second,
+		LongRunningFunc:       func(_ *http.Request, _ *request.RequestInfo) bool { return false },
+	}
+
+	h := DefaultBuildHandlerChain(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// confirm this is a no-op
+		if r.Context() != audit.WithAuditAnnotations(r.Context()) {
+			t.Error("unexpected double wrapping of context")
+		}
+
+		// confirm that we have an audit event
+		ae := request.AuditEventFrom(r.Context())
+		if ae == nil {
+			t.Error("unexpected nil audit event")
+		}
+
+		// confirm that the direct way of setting audit annotations later in the chain works as expected
+		audit.LogAnnotation(ae, "snorlax", "is cool too")
+
+		// confirm that the indirect way of setting audit annotations later in the chain also works
+		audit.AddAuditAnnotation(r.Context(), "dogs", "are okay")
+
+		_, _ = w.Write([]byte("done"))
+	}), c)
+	w := &testResponseWriter{header: http.Header{}}
+
+	h.ServeHTTP(w, &http.Request{URL: &url.URL{}})
+
+	if ok := w.respCode == http.StatusOK && w.writeHeaderCalled && string(w.data) == "done" && len(w.header.Get(auditinternal.HeaderAuditID)) > 0; !ok {
+		t.Errorf("invalid response: %#v", w)
+	}
+	if len(backend.events) == 0 {
+		t.Error("expected audit events, got none")
+	}
+	for _, event := range backend.events {
+		if want := map[string]string{"pandas": "are awesome", "snorlax": "is cool too", "dogs": "are okay"}; !reflect.DeepEqual(event.Annotations, want) {
+			t.Errorf("event has unexpected annotations: %#v, want annotations %#v", event, want)
+		}
```

Comment:
This will have a better failure message:
```go
want := map[string]string{"pandas": "are awesome", "snorlax": "is cool too", "dogs": "are okay"}
if diff := cmp.Diff(want, event.Annotations); diff != "" {
  t.Errorf("event has unexpected annotations (-want +got): %s", diff)
}
```